File: tools/logging/src/main/java/top/bettercode/summer/logging/ApplicationContextHolder.java
package top.bettercode.summer.logging;

import java.util.Locale;
import java.util.Optional;
import org.jetbrains.annotations.NotNull;
import org.springframework.beans.BeansException;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;
import org.springframework.context.NoSuchMessageException;
import org.springframework.core.convert.ConversionService;
import org.springframework.core.convert.support.DefaultConversionService;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;

/**
 * @author Peter Wu
 */
public class ApplicationContextHolder implements ApplicationContextAware {

  private static ApplicationContext applicationContext;

  @Override
  public void setApplicationContext(@NotNull ApplicationContext applicationContext)
      throws BeansException {
    ApplicationContextHolder.applicationContext = applicationContext;
  }

  public static ApplicationContext getApplicationContext() {
    return applicationContext;
  }

  public static <T> T getBean(String s, Class<T> aClass) throws BeansException {
    return applicationContext == null ? null : applicationContext.getBean(s, aClass);
  }

  public static <T> T getBean(Class<T> aClass) throws BeansException {
    return applicationContext == null ? null : applicationContext.getBean(aClass);
  }

  public static String getMessage(String s, Object[] objects, Locale locale)
      throws NoSuchMessageException {
    return applicationContext == null ? null : applicationContext.getMessage(s, objects, locale);
  }

  public static String getProperty(String key) {
    return applicationContext == null ? null : applicationContext.getEnvironment().getProperty(key);
  }


  public static String getProperty(String key, String defaultValue) {
    return applicationContext == null ? null
        : applicationContext.getEnvironment().getProperty(key, defaultValue);
  }


  public static <T> T getProperty(String key, Class<T> targetType) {
    return applicationContext == null ? null
        : applicationContext.getEnvironment().getProperty(key, targetType);
  }


  public static <T> T getProperty(String key, Class<T> targetType, T defaultValue) {
    return applicationContext == null ? null
        : applicationContext.getEnvironment().getProperty(key, targetType, defaultValue);
  }

  public static ConversionService getConversionService() {
    return applicationContext == null ? new DefaultConversionService()
        : applicationContext.getBean(ConversionService.class);
  }
  public static Optional<ServletRequestAttributes> getRequestAttributes() {
    ServletRequestAttributes requestAttributes = (ServletRequestAttributes) RequestContextHolder
        .getRequestAttributes();
    return Optional.ofNullable(requestAttributes);
  }
}

File: web/src/main/kotlin/top/bettercode/summer/web/support/ApplicationContextHolder.kt
package top.bettercode.summer.web.support

import org.springframework.context.ApplicationContext
import org.springframework.core.convert.ConversionService
import org.springframework.core.convert.support.DefaultConversionService
import org.springframework.web.context.request.RequestContextHolder
import org.springframework.web.context.request.ServletRequestAttributes
import java.util.*

/**
 * @author Peter Wu
 */
object ApplicationContextHolder {

    @JvmStatic
    var applicationContext: ApplicationContext? = null


    @JvmStatic
    fun <T> getBean(s: String, clazz: Class<T>): T? {
        return applicationContext?.getBean(s, clazz)
    }

    @JvmStatic
    fun <T> getBean(clazz: Class<T>): T? {
        return applicationContext?.getBean(clazz)
    }

    @JvmStatic
    fun getMessage(s: String, objects: Array<Any?>?, locale: Locale): String? {
        return applicationContext?.getMessage(s, objects, locale)
    }

    @JvmStatic
    fun getProperty(key: String): String? {
        return applicationContext?.environment?.getProperty(key)
    }

    @JvmStatic
    fun getProperty(key: String, defaultValue: String): String? {
        return applicationContext?.environment?.getProperty(key, defaultValue)
    }

    @JvmStatic
    fun <T> getProperty(key: String, targetType: Class<T>): T? {
        return applicationContext?.environment?.getProperty(key, targetType)
    }

    @JvmStatic
    fun <T> getProperty(key: String, targetType: Class<T>, defaultValue: T): T? {
        return applicationContext?.environment?.getProperty(key, targetType, defaultValue!!)
    }

    @JvmStatic
    val activeProfiles: Array<String>
        get() = applicationContext?.environment?.activeProfiles ?: emptyArray()

    @JvmStatic
    val isTest: Boolean
        get() = activeProfiles.any { it.contains("test") }

    @JvmStatic
    val isDev: Boolean
        get() = activeProfiles.any { it.contains("dev") }

    @JvmStatic
    val conversionService: ConversionService
        get() = applicationContext?.getBean(ConversionService::class.java)
                ?: DefaultConversionService()

    @JvmStatic
    val requestAttributes: Optional<ServletRequestAttributes>
        get() {
            val requestAttributes = RequestContextHolder
                    .getRequestAttributes() as ServletRequestAttributes
            return Optional.ofNullable(requestAttributes)
        }

}
